[quant] Add support for 2D indices for quantized embedding operators

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47766 [quant] Add support for 2D indices for quantized embedding operators**

Summary:
The operator now supports accepting 2D indices as inputs.
For embedding operators, we set the default offsets in the op since the FBGEMM kernel expects it to be set
Output shape depends on the shape if the indices.

For embedding_bag operator, if indices is 2D (B, N) then offsets should be set to None by user. In this case
the input is interpreted as B bags each of fixed length N. Output shape is still 2-D in this case.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingOps.test_embedding_bag_2d_indices
python test/test_quantization.py TestQuantizedEmbeddingOps.test_embedding_2d_indices

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24895048](https://our.internmc.facebook.com/intern/diff/D24895048)

supriyar committed:
[quant] Add support for 2D indices for quantized embedding operators

Summary:
The operator now supports accepting 2D indices as inputs.
For embedding operators, we set the default offsets in the op since the FBGEMM kernel expects it to be set
Output shape depends on the shape if the indices.

For embedding_bag operator, if indices is 2D (B, N) then offsets should be set to None by user. In this case
the input is interpreted as B bags each of fixed length N. Output shape is still 2-D in this case.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingOps.test_embedding_bag_2d_indices
python test/test_quantization.py TestQuantizedEmbeddingOps.test_embedding_2d_indices

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qembeddingbag.cpp:
```diff
@@ -237,9 +236,13 @@ at::Tensor embedding_bag_byte_impl(
-  auto output = at::empty(shape, weight.options().dtype(at::kFloat));
+  if (indices.dim() == 2 && is_embedding_op) {
+    std::vector<int64_t> shape = {indices.size(0), indices.size(1), D};
+    output = at::empty(shape, weight.options().dtype(at::kFloat));
```
nit: we can move output to outside of the branch and declare `std::vector<int64_t> shape` outside of the branch instead.

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qembeddingbag.cpp:
```diff
@@ -563,12 +605,20 @@ class QEmbedding final {
+    // Set default offsets here since the FBGEMM lookup op expects it.
     const auto offsets_size = indices.numel();
-    at::Tensor offsets = at::arange(0, offsets_size, at::kLong);
+    at::Tensor offsets = at::arange(0, offsets_size, indices.scalar_type());
```
question: do we support other types of indices?

supriyar replied:
You mean dtype? We support int32, int64 types

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qembeddingbag.cpp:
```diff
@@ -563,12 +605,20 @@ class QEmbedding final {
-      return packed_weight->embeddingbag_byte(
-          indices, offsets, pruned_weights, c10::nullopt, c10::nullopt, false);
+      return embedding_bag_byte_helper(
+          packed_weight->get_packed_weight().contiguous(),
```
I think the common practice is to define embeddingbag_byte on packed_weight since packed_weight can be in different backends (fbgemm/qnnpack), examples can be found in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qconv.cpp. why do we change this?

supriyar replied:
This can be changed. The main motivation was to not have to change the definition of `embeddingbag_byte` operator to add a new bool argument (but it is doable). 
In this case the packed weight is still stored as a PyTorch tensor so it can be directly accessed.

jerryzh168 replied:
I see, if you need to distinguish between embedding/emedding_bag in the invocation to packed weight, I think you can follow ConvPackedParams as well, we can add a new template argument, (True/False) maybe to distinguish between these two cases. there might be implications for serialization if we change this though.

supriyar replied:
i think for now I can add a bool argument, since template might make it too complex for this PR.

supriyar committed:
Update on "[quant] Add support for 2D indices for quantized embedding operators"


Summary:
The operator now supports accepting 2D indices as inputs.
For embedding operators, we set the default offsets in the op since the FBGEMM kernel expects it to be set
Output shape depends on the shape if the indices.

For embedding_bag operator, if indices is 2D (B, N) then offsets should be set to None by user. In this case
the input is interpreted as B bags each of fixed length N. Output shape is still 2-D in this case.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingOps.test_embedding_bag_2d_indices
python test/test_quantization.py TestQuantizedEmbeddingOps.test_embedding_2d_indices

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24895048](https://our.internmc.facebook.com/intern/diff/D24895048)

[ghstack-poisoned]

jerryzh168 reviewed (approved):
looks good, thanks!

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qembeddingbag.cpp:
```diff
@@ -497,7 +540,8 @@ Tensor embedding_bag_byte_rowwise_offsets(
       compressed_indices_mapping,
-      include_last_offset);
+      include_last_offset,
+      false);
```
nit: please add a comment for the false:
`false /* is_embedding_op */`

jerryzh168 replied:
same for other occurrences

supriyar committed:
Update on "[quant] Add support for 2D indices for quantized embedding operators"


Summary:
The operator now supports accepting 2D indices as inputs.
For embedding operators, we set the default offsets in the op since the FBGEMM kernel expects it to be set
Output shape depends on the shape if the indices.

For embedding_bag operator, if indices is 2D (B, N) then offsets should be set to None by user. In this case
the input is interpreted as B bags each of fixed length N. Output shape is still 2-D in this case.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingOps.test_embedding_bag_2d_indices
python test/test_quantization.py TestQuantizedEmbeddingOps.test_embedding_2d_indices

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24895048](https://our.internmc.facebook.com/intern/diff/D24895048)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add support for 2D indices for quantized embedding operators"


Summary:
The operator now supports accepting 2D indices as inputs.
For embedding operators, we set the default offsets in the op since the FBGEMM kernel expects it to be set
Output shape depends on the shape if the indices.

For embedding_bag operator, if indices is 2D (B, N) then offsets should be set to None by user. In this case
the input is interpreted as B bags each of fixed length N. Output shape is still 2-D in this case.

Test Plan:
python test/test_quantization.py TestQuantizedEmbeddingOps.test_embedding_bag_2d_indices
python test/test_quantization.py TestQuantizedEmbeddingOps.test_embedding_2d_indices

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24895048](https://our.internmc.facebook.com/intern/diff/D24895048)

[ghstack-poisoned]
